File: c/sort/selection_sort.c
/*
 * Title    : Selection Sort
 * Problem  : Sort N integers in ascending order
 * Category : sorting
 * Language : C
 * Author   : YeongWoong Kim
 * note     : Reads N(1~100) integers, sorts them in ascending order, and prints the result.
 */

#include <stdio.h>
#define MAX_SIZE 100

// Function prototype
void sort(int array[], int size);
void print_array(int array[], int size);

int main() {
    int size;

    // Ask the user for the number of elements
    printf("Enter the number of elements to sort(1~100): ");
    scanf_s("%d", &size);

    // Validate size
    if (size < 1 || size > MAX_SIZE) {
        printf("Invalid size.\n");
        return 1;
    }

    int array[MAX_SIZE];

    // Input elements from the user
    printf("Enter %d integers:\n", size);
    for (int i = 0; i < size; i++) {
        scanf_s("%d", &array[i]);
    }

    // Print before sorting
    printf("\nBefore sort:\n");
    print_array(array, size);

    // Call selection sort function
    sort(array, size);

    // Print sorted result
    printf("\nSorted result:\n");
    print_array(array, size);

    return 0;
}

// Selection sort function
void sort(int a[], int size) {
    int i, j, min, position, temp;

    // Loop for each position in the array
    for (j = 0; j < size; j++) {
        min = a[j];
        position = j;

        // Find the minimum value in the remaining array
        for (i = j; i < size; i++) {
            if (a[i] < min) {
                min = a[i];
                position = i;
            }
        }

        // Swap the found minimum with the current position
        temp = a[position];
        a[position] = a[j];
        a[j] = temp;
    }
}

void print_array(int a[], int size) {
    for (int i = 0; i < size; i++) {
        printf("%d\t", a[i]);
        if (i != 0 && (i + 1) % 10 == 0)
            printf("\n");
    }
    printf("\n");
}

File: java/sort/selection_sort.java
/*
 * Title    : Selection Sort
 * Problem  : Sort N integers in ascending order
 * Category : sorting
 * Language : Java
 * Author   : Yeonjun Cho
 * Note     : Reads N(1~100) integers, sorts them in ascending order, and prints the result.
 */

import java.util.Scanner;

public class SelectionSort {
    static final int MAX_SIZE = 100;

    // Selection sort
    static void sort(int[] array, int size) {
        for (int j = 0; j < size; j++) {
            int min = array[j];
            int pos = j;

            for (int i = j; i < size; i++) {
                if (array[i] < min) {
                    min = array[i];
                    pos = i;
                }
            }

            int tmp = array[pos];
            array[pos] = array[j];
            array[j] = tmp;
        }
    }

    static void printArray(int[] array, int size) {
        for (int i = 0; i < size; i++) {
            System.out.print(array[i] + "\t");
            if (i != 0 && (i + 1) % 10 == 0) System.out.println();
        }
        System.out.println();
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.print("Enter the number of elements to sort(1~100): ");
        int size = sc.nextInt();

        if (size < 1 || size > MAX_SIZE) {
            System.out.println("Invalid size.");
            return;
        }

        // Allocate exactly as many elements as needed (saves memory)
        int[] array = new int[size];

        System.out.println("Enter " + size + " integers:");
        for (int i = 0; i < size; i++) {
            array[i] = sc.nextInt();
        }

        System.out.println("\nBefore sort:");
        printArray(array, size);

        sort(array, size);

        System.out.println("\nSorted result:");
        printArray(array, size);

        sc.close();
    }
}
